### PR TITLE
Increase tracking integration to 20ms.

### PR DESCRIPTION
**Pull request by gsmcmullin**

### CODE REVIEW

**Comment by gsmcmullin** [comment]
retest this please

**Comment by gsmcmullin** [comment]
This is working on my desk, but lets wait for HITL results...

**Comment by fnoble** [comment]
Can you rebase? Sorry

File: src/track.c [hunk 1]
```diff
@@ -270,15 +303,39 @@ void tracking_channel_update(u8 channel)
...
+          log_warn("False phase lock detect PRN%d: err=%f\n", chan->prn+1, err);
+          chan->tl_state.carr_freq += err;
+          chan->tl_state.carr_filt.y = chan->tl_state.carr_freq;
+        }
+      }
+
+      if ((chan->TOW_ms > 0) && (chan->int_ms == 1) &&
+          (chan->nav_msg.bit_phase == chan->nav_msg.bit_phase_ref)) {
+        /* Now that we have TOW we can transition to longer integration */
+        printf("Increasing integration time for PRN %d\n", chan->prn+1);
```
Comment (fnoble) [inline src/track.c]: this should be a `log_info`

**Comment by gsmcmullin** [comment]
retest this please

**Comment by cbeighley** [comment]
retest this please

**Comment by cbeighley** [comment]
retest this please

**Comment by cbeighley** [comment]
HITL testing looks good

**Comment by cbeighley** [comment]
@fnoble @henryhallam can you review

**Comment by henryhallam** [comment]
Update the required NAP firmware version in main.c

File: src/manage.c [hunk 2]
```diff
@@ -370,10 +370,25 @@ static msg_t manage_track_thread(void *arg)
   return 0;
 }
 
+static u16 iq_output_mask;
```
Comment (fnoble) [inline src/manage.c]: Explicitly initialize to zero

File: src/track.c [hunk 3]
```diff
@@ -261,15 +294,41 @@ void tracking_channel_update(u8 channel)
       chan->code_phase_rate_fp = chan->code_phase_rate
         * NAP_TRACK_CODE_PHASE_RATE_UNITS_PER_HZ;
 
-      chan->carrier_freq_fp_prev = chan->carrier_freq_fp;
       chan->carrier_freq_fp = chan->carrier_freq
         * NAP_TRACK_CARRIER_FREQ_UNITS_PER_HZ;
 
+      if (chan->int_ms != 1) {
+        s32 I = (cs[1].I - chan->alias_detect.first_I) / (LONG_INTEGRATION_INTERVAL-1);
+        s32 Q = (cs[1].Q - chan->alias_detect.first_Q) / (LONG_INTEGRATION_INTERVAL-1);
```
Comment (fnoble) [inline src/track.c]: Should this be moved into the alias detection function?

File: src/track.c [hunk 4]
```diff
@@ -203,23 +221,47 @@ void tracking_channel_update(u8 channel)
...
-        chan->TOW_ms++;
-        if (chan->TOW_ms == 7*24*60*60*1000)
-          chan->TOW_ms = 0;
+        chan->TOW_ms += chan->short_cycle ? 1 : (chan->int_ms-1);
+        chan->TOW_ms %= 7*24*60*60*1000;
+      }
 
-        /* Turn off FLL aiding. For now we do this here because having a valid
-         * TOW is a very good indication that the tracking loops have locked. */
-        chan->tl_state.carr_filt.aiding_igain = 0;
```
Comment (henryhallam) [inline src/track.c]: So now we're always using FLL aiding?

File: src/track.c [hunk 5]
```diff
@@ -203,23 +221,47 @@ void tracking_channel_update(u8 channel)
...
+      if (chan->int_ms > 1) {
+        /* If we're doing long integrations alternate between short and long
+         * cycles.
```
Comment (henryhallam) [inline src/track.c]: Explain why (FPGA update latency)

File: src/track.c (hunk 4 above)
Comment (gsmcmullin) [inline src/track.c]: No.  When we switch to 20ms integration the loop filter is reinitialised with an FLL gain of 0.  The PLL filter coefficients also need to be recalculated, to aided_tl_init() is called with the new parameters.

File: src/track.c [hunk 6]
```diff
@@ -261,15 +294,41 @@ void tracking_channel_update(u8 channel)
...
+      if ((chan->TOW_ms > 0) && (chan->int_ms == 1) &&
+          (chan->nav_msg.bit_phase == chan->nav_msg.bit_phase_ref)) {
+        /* Now that we have TOW we can transition to longer integration */
+        log_info("Increasing integration time for PRN %d\n", chan->prn+1);
+        chan->int_ms = LONG_INTEGRATION_INTERVAL;
+        chan->short_cycle = true;
+
+        cn0_est_init(&chan->cn0_est, 1e3/chan->int_ms, chan->cn0, 5, 1e3);
+
+        /* Recalculate filter coefficients */
```
Comment (henryhallam) [inline src/track.c]: Recalculate filter coefficients: now pure PLL

File: src/track.c (hunk 3 above)
Comment (gsmcmullin) [inline src/track.c]: I don't really think so.  This is a consequence of the fact that cs[] are the sum of the short and long interval accumulated in tracking_channel_get_corrs(). We could also keep extra state of the intermediate values around.  This is working, so I'd rather just leave it alone.  Either way, I think the alias detect functions in libswiftnav are fine.

**Comment by gsmcmullin** [comment]
Did the 0.11 NAP firmware get released?  Or is that coming with the 0.16 Piksi firmware?
I don't see it on downloads.swiftnav.com.

**Comment by cbeighley** [comment]
It'll be released in conjunction with Piksi FW v0.16.

**Comment by henryhallam** [comment]
Looks like there's no scope for transitioning back from 20ms + PLL mode to 1ms + FLL in the event that SNR drops off.  I guess that's OK, as long as we're sure that we get the nav bit sync right first time - once we're in 20ms mode we can no longer update our nav bit sync histogram.

**Comment by gsmcmullin** [comment]
That is correct.  It'll only transition once TOW has been decoded, so we have good confidence in our bit sync.

**Comment by fnoble** [comment]
:dancers:  YAAAY! :dancers:

**Pull request merged**